backend/src/permissions/permissions.guard.ts

```diff
@@ -0,0 +1,95 @@
+import { NotesService } from '../notes/notes.service';
+import { NotePermission } from './note-permission.enum';
+import { PermissionsService } from './permissions.service';
+import { permissionMetadataKey } from './require-permission.decorator';
+import { RequiredPermission } from './required-permission.enum';
+
+/**
+ * This guards controller methods from access, if the user has not the appropriate permissions.
+ * The permissions are set via the {@link Permissions} decorator in addition to this guard.
+ * If the check permission is not CREATE the method needs to extract the noteIdOrAlias from
+ * request.params['noteIdOrAlias'] or request.headers['hedgedoc-note'] to check if the user has the permission.
+ */
+@Injectable()
+export class PermissionsGuard implements CanActivate {
+  constructor(
+    private readonly logger: ConsoleLoggerService,
+    private reflector: Reflector,
+    private permissionsService: PermissionsService,
+    private noteService: NotesService,
+  ) {
+    this.logger.setContext(PermissionsGuard.name);
+  }
+
+  async canActivate(context: ExecutionContext): Promise<boolean> {
+    const requiredAccessLevel = this.extractRequiredAccessLevel(context);
+    if (requiredAccessLevel === undefined) {
+      return false;
+    }
+    const request: CompleteRequest = context.switchToHttp().getRequest();
+    const user = request.user ?? null;
+
+    // handle CREATE requiredAccessLevel, as this does not need any note
+    if (requiredAccessLevel === RequiredPermission.CREATE) {
+      return this.permissionsService.mayCreate(user);
+    }
+
+    // Get the note from the parameter noteIdOrAlias or the http header hedgedoc-note
+    // Attention: This gets the note an additional time if used in conjunction with GetNoteInterceptor or NoteHeaderInterceptor
+    const noteIdOrAlias = extractNoteIdOrAlias(request);
+    if (noteIdOrAlias === undefined) {
+      this.logger.error(
+        'Could not find noteIdOrAlias metadata. This should never happen. If you see this, please open an issue at https://github.com/hedgedoc/hedgedoc/issues',
+      );
+      return false;
+    }
+
+    const note = await this.noteService.getNoteByIdOrAlias(noteIdOrAlias);
+    return this.isNotePermissionFulfillingRequiredAccessLevel(
+      requiredAccessLevel,
+      await this.permissionsService.determinePermission(user, note),
+    );
+  }
+
+  private extractRequiredAccessLevel(
+    context: ExecutionContext,
+  ): RequiredPermission | undefined {
+    const requiredAccessLevel = this.reflector.get<RequiredPermission>(
+      permissionMetadataKey,
+      context.getHandler(),
+    );
+    // If no requiredAccessLevel are set this is probably an error and this guard should not let the request pass
+    if (!requiredAccessLevel) {
+      this.logger.error(
+        'Could not find requiredAccessLevel metadata. This should never happen. If you see this, please open an issue at https://github.com/hedgedoc/hedgedoc/issues',
```

Comment:
Maybe we want to extract the "This should never happen"-text as a variable for easy changes in the future?

Reply:
Huh sounds like a good idea